File: Demo/MNG-Ads-SDK/GADBlueStackMediationAdapter.m
//
//  GADBlueStackMediationAdapter.m
//  MNG-Ads-SDK-Demo
//
//  Created by anypli on 8/3/2022.
//  Copyright © 2022 Bensalah Med Amine. All rights reserved.
//

#import "GADBlueStackMediationAdapter.h"
#import "GADBlueStackBannerRenderer.h"
#import "GADBlueStackInterstitialRenderer.h"
#import "GADBlueStackNativeAdRenderer.h"
#import "GADBlueStackRewardedRenderer.h"

#import "BlueStackGADAdNetworkExtras.h"

@interface GADBlueStackMediationAdapter () <GADMediationAdapter>

@end

@implementation GADBlueStackMediationAdapter{
  
    /// Bluestack  Network interstitial ad wrapper.
    GADBlueStackInterstitialRenderer *_interstitial;
    /// Bluestack  Network banner ad wrapper.
    GADBlueStackBannerRenderer *_banner;
    /// Bluestack  Network banner ad wrapper.
    GADBlueStackNativeAdRenderer *_native;
    /// BluestackNetwork rewarded ad wrapper.
    GADBlueStackRewardedRenderer *_rewardedAd;
}
UIViewController* _viewController;

-(id)init {
    self = [super init];
    return self;
}

+ (GADVersionNumber)adSDKVersion {
    GADVersionNumber version = {0};
      NSArray<NSString *> *components = [@"4.0.0" componentsSeparatedByString:@"."];
      if (components.count == 4) {
        version.majorVersion = components[0].integerValue;
        version.minorVersion = components[1].integerValue;
        version.patchVersion = components[2].integerValue * 100 + components[3].integerValue;
      }
      return version;
    
}


+ (GADVersionNumber)adapterVersion {
    GADVersionNumber version = {0};
      NSArray<NSString *> *components = [@"4.0.0" componentsSeparatedByString:@"."];
      if (components.count == 4) {
        version.majorVersion = components[0].integerValue;
        version.minorVersion = components[1].integerValue;
        version.patchVersion = components[2].integerValue * 100 + components[3].integerValue;
      }
      return version;
}


+ (nullable Class<GADAdNetworkExtras>)networkExtrasClass {
    return [BlueStackGADAdNetworkExtras class];
}

- (void)loadBannerForAdConfiguration:(GADMediationBannerAdConfiguration *)adConfiguration
                   completionHandler:(GADMediationBannerLoadCompletionHandler)completionHandler {
    
    
    _banner = [[GADBlueStackBannerRenderer alloc] init];
    [_banner renderBannerForAdConfiguration:adConfiguration completionHandler:completionHandler];
}

- (void)loadInterstitialForAdConfiguration:
(GADMediationInterstitialAdConfiguration *)adConfiguration
                         completionHandler:
(GADMediationInterstitialLoadCompletionHandler)completionHandler {

      _interstitial = [[GADBlueStackInterstitialRenderer alloc] init];
      [_interstitial setViewController:_viewController];
      [_interstitial renderInterstitialForAdConfiguration:adConfiguration
                                        completionHandler:completionHandler];
}


- (void)loadNativeAdForAdConfiguration:(GADMediationNativeAdConfiguration *)adConfiguration
                     completionHandler:(GADMediationNativeLoadCompletionHandler)completionHandler {
   
      _native = [[GADBlueStackNativeAdRenderer alloc] init];
      [_native setViewController:_viewController];
      [_native renderNativeAdForAdConfiguration:adConfiguration completionHandler:completionHandler];
}

- (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)adConfiguration
                       completionHandler:
                           (GADMediationRewardedLoadCompletionHandler)completionHandler {

    _rewardedAd = [[GADBlueStackRewardedRenderer alloc] init];
    [_rewardedAd loadRewardedAdForAdConfiguration:adConfiguration
                                completionHandler:completionHandler];
    
}

/**
 viewController that th ad will be shown
@warning required in interstitial
*/
+(void)setViewController:(UIViewController*)viewController{
    _viewController = viewController;
}

@end

File: DemoSwift/DemoSwift/GoogleMobileAds-Adapter/Adapter/GADBlueStackMediationAdapter.swift
//
//  GADBlueStackMediationAdapter.swift
//  BlueStackDemoDev
//
//  Created by MedSghaier on 18/8/2022.
//

import Foundation
import GoogleMobileAds
import UIKit

//the class name must be  appName.className (BlueStackDemoDev.GADBlueStackMediationAdapter)
//See this https://developers.google.com/ad-manager/mobile-ads-sdk/ios/custom-events/interstitial //className section
class GADBlueStackMediationAdapter : NSObject {
    
    let viewController : UIViewController
    let _banner  : GADBlueStackBannerRenderer = GADBlueStackBannerRenderer()
    let _interstitial  : GADBlueStackInterstitialRenderer = GADBlueStackInterstitialRenderer()
    let _native : GADBlueStackNativeAdRenderer = GADBlueStackNativeAdRenderer()
    
    init(viewController : UIViewController) {
        self.viewController = viewController
        super.init()
    }
    
    
    //Version
    //MARK: load banner
    func loadBannerForAdConfiguration(adConfiguration : GADMediationBannerAdConfiguration, completionHandler : @escaping GADMediationBannerLoadCompletionHandler) {
        
        _banner.renderBannerForAdConfiguration(adConfiguration: adConfiguration, completionHandler: completionHandler)
        
    }
    
    //MARK: load Interstitial
    func loadInterstitialForAdConfiguration(adConfiguration : GADMediationInterstitialAdConfiguration, completionHandler : @escaping GADMediationInterstitialLoadCompletionHandler) {
        
        _interstitial.viewController = viewController
        _interstitial.renderInterstitialForAdConfiguration(adConfiguration: adConfiguration, completionHandler: completionHandler)

    }
    
    //MARK: load NativeAd
    func loadNativeAdForAdConfiguration(adConfiguration : GADMediationNativeAdConfiguration, completionHandler : @escaping GADMediationNativeLoadCompletionHandler) {
        
        _native.viewController = viewController
        _native.renderNativeAdForAdConfiguration(adConfiguration: adConfiguration, completionHandler: completionHandler)

    }
    
    
    
}
